katdal/sensordata.py

```diff
@@ -462,6 +462,73 @@ def _safe_linear_interp(xi, yi, x):
     return (1.0 - end_weight) * yi[start] + end_weight * yi[end]
 
 
+_KATCP_DECODER = {'integer': int, 'float': float, 'boolean': lambda x: x == '1',
+                  'lru': str, 'timestamp': float, 'discrete': str,
+                  'address': str, 'string': str}
+
+
+def get_sensor_from_katstore(store, name, start_time, end_time):
+    """Get raw sensor data from katstore (CAM's central sensor database).
+
+    Parameters
+    ----------
+    store : string
+        Hostname / endpoint of katstore webserver speaking katstore REST API
+    name : string
+        Sensor name (the escaped version with underscores)
+    start_time, end_time : float
+        Time range for sensor records as UTC seconds since Unix epoch
+
+    Returns
+    -------
+    data : :class:`RecordSensorData` object
+        Retrieved sensor data with 'timestamp', 'value' and 'status' fields
+
+    Raises
+    ------
+    ConnectionError
+        If this cannot connect to the katstore server
+    ValueError
+        If connection succeeded but interaction with REST API failed
+    KeyError
+        If the sensor was not found in the store or it has no data
+    """
+    # First check existence of sensor and get its KATCP type
+    url = "http://{}/katstore/sensors".format(store)
+    try:
+        result = requests.get(url, params={'sensors': name})
+    except requests.exceptions.ConnectionError as exc:
+        raise_from(ConnectionError("Could not connect to sensor store '%s'" %
+                                   (store,)), exc)
+    try:
+        sensor_info = {rec[0]: rec[2] for rec in result.json()}
+    except ValueError as exc:
+        raise_from(ValueError("Could not retrieve sensor info from '%s' (%d: %s)" %
+                              (url, result.status_code, result.reason)), exc)
+    if name not in sensor_info:
+        raise KeyError("Sensor store has no sensor named '%s'" % (name,))
+    decode = _KATCP_DECODER[sensor_info[name]['type']]
+    # Now get the historical data for the sensor in the given time range
+    url = "http://{}/katstore/samples".format(store)
+    params = {'sensor': name, 'start': start_time, 'end': end_time,
+              'time_type': 's', 'limit': 1000000}
+    try:
+        result = requests.get(url, params)
+    except requests.exceptions.ConnectionError as exc:
+        raise_from(ConnectionError("Could not connect to sensor store '%s'" %
+                                   (store,)), exc)
+    try:
+        samples = [(rec[1], decode(rec[3]), rec[5])
+                   for rec in result.json() if rec[4] == name]
+    except ValueError as exc:
+        raise_from(ValueError("Could not retrieve samples from '%s' (%d: %s)" %
+                              (url, result.status_code, result.reason)), exc)
```

Comment:
Again, check the status_code before even trying to parse the response.

Reply:
As before (the JSON error covers all cases).

Reply:
Does the API guarantee to return a non-JSON payload in the case of an error?

Reply:
@ajoubertza can hopefully clarify this. That has been my experience with simple 404's and misspelt sensors (an empty JSON for the latter).

Reply:
No guarantees.  This interface is not for public consumption.

Reply:
I'm thinking I should even remove the `sensor_store` parameter from `VisDataV4` since I want to emphasise that this is a power-user / debug-level feature. On the other hand, overriding the sensor store during `katdal.open` would allow katdal to open a severely broken RDB file, so I'm probably keeping it there for now.

I do like the following though 😄 

```Python
f = katdal.open('blah')
f.sensor.store = 'portal.mkat.karoo.kat.ac.za'
```

That also works for H5 files 😉